Internalerror  KeyError: <class '_io.BufferedReader'>

Not sure what's exactly going on but pytest-xdist started emitting INTERNALERRORs and failing in version 1.23.1 (and 1.23.2). pytest-xdist-1.23.0 does not emit these.

**EDIT**: minimal repro:
```
$ cat test_foo.py
def test_foo():
    open('/dev/null', 'rb')
$ python3.6 -mpytest test_foo.py -W once -n2
```

To reproduce:
```
$ python3.6 -mvenv tmp
$ ./tmp/bin/python3.6 -mpip install pytest pytest-xdist six
$ git clone https://github.com/airspeed-velocity/asv.git
$ cd asv
$ git checkout 6f729bb9808116c1ebaa77c6684a4c8120926051
$ ../tmp/bin/python3.6 -mpytest test/test_dev.py -n2
===================================== test session starts =====================================
platform linux -- Python 3.6.6, pytest-3.8.1, py-1.6.0, pluggy-0.7.1
rootdir: /home/pauli/tmp/asv, inifile: setup.cfg
plugins: xdist-1.23.2, forked-0.2
gw0 [6] / gw1 [6]
scheduling tests via LoadScheduling
.INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1383, in _save
INTERNALERROR>     dispatch = self._dispatch[tp]
INTERNALERROR> KeyError: <class '_io.BufferedReader'>
INTERNALERROR> 
INTERNALERROR> During handling of the above exception, another exception occurred:
INTERNALERROR> 
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/_pytest/main.py", line 178, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/_pytest/main.py", line 215, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "<remote exec>", line 65, in pytest_runtestloop
INTERNALERROR>   File "<remote exec>", line 82, in run_one_test
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 196, in _multicall
INTERNALERROR>     gen.send(outcome)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/_pytest/warnings.py", line 137, in pytest_runtest_protocol
INTERNALERROR>     yield
INTERNALERROR>   File "/usr/lib64/python3.6/contextlib.py", line 88, in __exit__
INTERNALERROR>     next(self.gen)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/_pytest/warnings.py", line 96, in catch_warnings_for_item
INTERNALERROR>     kwargs=dict(warning_message=warning_message, when=when, item=item)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/hooks.py", line 280, in call_historic
INTERNALERROR>     res = self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "<remote exec>", line 135, in pytest_warning_captured
INTERNALERROR>   File "<remote exec>", line 29, in sendevent
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 717, in send
INTERNALERROR>     self.gateway._send(Message.CHANNEL_DATA, self.id, dumps_internal(item))
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1354, in dumps_internal
INTERNALERROR>     return _Serializer().save(obj)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1372, in save
INTERNALERROR>     self._save(obj)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1390, in _save
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1475, in save_tuple
INTERNALERROR>     self._save(item)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1390, in _save
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1471, in save_dict
INTERNALERROR>     self._write_setitem(key, value)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1465, in _write_setitem
INTERNALERROR>     self._save(value)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1390, in _save
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1471, in save_dict
INTERNALERROR>     self._write_setitem(key, value)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1465, in _write_setitem
INTERNALERROR>     self._save(value)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/execnet/gateway_base.py", line 1388, in _save
INTERNALERROR>     raise DumpError("can't serialize %s" % (tp,))
INTERNALERROR> execnet.gateway_base.DumpError: can't serialize <class '_io.BufferedReader'>
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/_pytest/main.py", line 178, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/_pytest/main.py", line 215, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/hooks.py", line 258, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 67, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/manager.py", line 61, in <lambda>
INTERNALERROR>     firstresult=hook.spec_opts.get('firstresult'),
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 201, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 76, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/pluggy/callers.py", line 180, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/xdist/dsession.py", line 115, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/xdist/dsession.py", line 138, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "/home/pauli/tmp/tmp/lib64/python3.6/site-packages/xdist/dsession.py", line 180, in worker_workerfinished
INTERNALERROR>     assert not crashitem, (crashitem, node)
INTERNALERROR> AssertionError: ('test/test_dev.py::test_dev_with_repo_subdir', <WorkerController gw1>)
INTERNALERROR> assert not 'test/test_dev.py::test_dev_with_repo_subdir'

================================== 1 passed in 6.23 seconds ===================================
```

I have the similar problem at 1.23.0+ version. At 1.23.0 work correcly.


```python
platform win32 -- Python 3.6.6, pytest-3.9.3, py-1.7.0, pluggy-0.8.0 -- c:\service_test\.venv\scripts\python.exe
cachedir: .pytest_cache
rootdir: c:\service_test, inifile: pytest.ini
plugins: xdist-1.23.2, timeout-1.3.2, repeat-0.7.0, forked-0.2, allure-adaptor-1.7.10
timeout: 60.0s
timeout method: thread
timeout func_only: False
[gw0] win32 Python 3.6.6 cwd: c:\service_test
[gw1] win32 Python 3.6.6 cwd: c:\service_test
[gw0] Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 27 2018, 02:47:15) [MSC v.1900 32 bit (Intel)]
[gw1] Python 3.6.6 (v3.6.6:4cf1f54eb7, Jun 27 2018, 02:47:15) [MSC v.1900 32 bit (Intel)]
gw0 [27] / gw1 [27]
scheduling tests via LoadScheduling

tests/test_.py::TestArbitrmanagers::test__
tests/test_arbitrmanagers/test_arbitrmanagers.py::TestArbitrmanagers::test_am_with_valid_sro
[gw0] [  3%] PASSED tests/test_.py::TestArbitrmanagers::test__
tests/test_arbitrmanagers/test_arbitrmanagers.py::TestArbitrmanagers::test_am_inn
[gw1] [  7%] PASSED tests/test_arbitrmanagers/test_arbitrmanagers.py::TestArbitrmanagers::test_am_with_valid_sro
[gw0] [ 11%] PASSED tests/test_arbitrmanagers/test_arbitrmanagers.py::TestArbitrmanagers::test_am_inn INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1383, in _save
INTERNALERROR>     dispatch = self._dispatch[tp]
INTERNALERROR> KeyError: <class 'sqlalchemy.util.langhelpers._hash_limit_string'>
INTERNALERROR>
INTERNALERROR> During handling of the above exception, another exception occurred:
INTERNALERROR>
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\main.py", line 184, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\main.py", line 224, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\hooks.py", line 284, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1383, in _save
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>     dispatch = self._dispatch[tp]
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR> KeyError: <class 'sqlalchemy.util.langhelpers._hash_limit_string'>
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>
INTERNALERROR> During handling of the above exception, another exception occurred:
INTERNALERROR>
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\main.py", line 184, in wrap_session
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "<remote exec>", line 65, in pytest_runtestloop
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\main.py", line 224, in _main
INTERNALERROR>   File "<remote exec>", line 82, in run_one_test
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\hooks.py", line 284, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\hooks.py", line 284, in __call__
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 203, in _multicall
INTERNALERROR>     gen.send(outcome)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\warnings.py", line 138, in pytest_runtest_protocol
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>     yield
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR>   File "c:\python36\Lib\contextlib.py", line 88, in __exit__
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>     next(self.gen)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\warnings.py", line 97, in catch_warnings_for_item
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>     kwargs=dict(warning_message=warning_message, when=when, item=item)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\hooks.py", line 306, in call_historic
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>     res = self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "<remote exec>", line 65, in pytest_runtestloop
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
INTERNALERROR>   File "<remote exec>", line 82, in run_one_test
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\hooks.py", line 284, in __call__
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 203, in _multicall
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>     gen.send(outcome)
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\warnings.py", line 138, in pytest_runtest_protocol
INTERNALERROR>   File "<remote exec>", line 135, in pytest_warning_captured
INTERNALERROR>     yield
INTERNALERROR>   File "<remote exec>", line 29, in sendevent
INTERNALERROR>   File "c:\python36\Lib\contextlib.py", line 88, in __exit__
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 717, in send
INTERNALERROR>     next(self.gen)
INTERNALERROR>     self.gateway._send(Message.CHANNEL_DATA, self.id, dumps_internal(item))
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\warnings.py", line 97, in catch_warnings_for_item
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1354, in dumps_internal
INTERNALERROR>     kwargs=dict(warning_message=warning_message, when=when, item=item)
INTERNALERROR>     return _Serializer().save(obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\hooks.py", line 306, in call_historic
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1372, in save
INTERNALERROR>     res = self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>     self._save(obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1475, in save_tuple
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>     self._save(item)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1471, in save_dict
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>     self._write_setitem(key, value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1465, in _write_setitem
INTERNALERROR>     self._save(value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "<remote exec>", line 135, in pytest_warning_captured
INTERNALERROR>   File "<remote exec>", line 29, in sendevent
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 717, in send
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1471, in save_dict
INTERNALERROR>     self.gateway._send(Message.CHANNEL_DATA, self.id, dumps_internal(item))
INTERNALERROR>     self._write_setitem(key, value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1354, in dumps_internal
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1465, in _write_setitem
INTERNALERROR>     return _Serializer().save(obj)
INTERNALERROR>     self._save(value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1372, in save
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>     self._save(obj)
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1475, in save_tuple
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>     self._save(item)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1475, in save_tuple
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1388, in _save
INTERNALERROR>     self._save(item)
INTERNALERROR>     raise DumpError("can't serialize %s" % (tp,))
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1471, in save_dict
INTERNALERROR> execnet.gateway_base.DumpError: can't serialize <class 'sqlalchemy.util.langhelpers._hash_limit_string'>
INTERNALERROR>     self._write_setitem(key, value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1465, in _write_setitem
INTERNALERROR>     self._save(value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1471, in save_dict
INTERNALERROR>     self._write_setitem(key, value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1465, in _write_setitem
INTERNALERROR>     self._save(value)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1390, in _save
INTERNALERROR>     dispatch(self, obj)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1475, in save_tuple
INTERNALERROR>     self._save(item)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\execnet\gateway_base.py", line 1388, in _save
INTERNALERROR>     raise DumpError("can't serialize %s" % (tp,))
INTERNALERROR> execnet.gateway_base.DumpError: can't serialize <class 'sqlalchemy.util.langhelpers._hash_limit_string'>

INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\main.py", line 184, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\_pytest\main.py", line 224, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\hooks.py", line 284, in __call__
INTERNALERROR>     return self._hookexec(self, self.get_hookimpls(), kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
INTERNALERROR>     firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 208, in _multicall
INTERNALERROR>     return outcome.get_result()
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 80, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\pluggy\callers.py", line 187, in _multicall
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\xdist\dsession.py", line 115, in pytest_runtestloop
INTERNALERROR>     self.loop_once()
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\xdist\dsession.py", line 138, in loop_once
INTERNALERROR>     call(**kwargs)
INTERNALERROR>   File "c:\service_test\.venv\lib\site-packages\xdist\dsession.py", line 180, in worker_workerfinished
INTERNALERROR>     assert not crashitem, (crashitem, node)
INTERNALERROR> AssertionError: ('tests/test_arbitrmanagers/test_arbitrmanagers.py::TestArbitrmanagers::()::test_am_inn', <WorkerController gw0>)
INTERNALERROR> assert not 'tests/test_arbitrmanagers/test_arbitrmanagers.py::TestArbitrmanagers::()::test_am_inn'
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -21,23 +21,26 @@ env:
 - TOXENV=py-pytest31
 - TOXENV=py-pytest32
 - TOXENV=py-pytest33
-- TOXENV=py-pytestlatest
 
 install: pip install tox setuptools_scm
 script: tox
 
 stages:
-- linting
+- baseline
 - test
 - name: deploy
   if: repo = pytest-dev/pytest-xdist AND tag IS present
 
 jobs:
   include:
-    - stage: linting
+    - stage: baseline
       python: '3.6'
-      script:
-      - tox -e linting
+      env: TOXENV=linting
+    - python: '3.6'
+      env: TOXENV=py36-pytestlatest
+    - python: '2.7'
+      env: TOXENV=py27-pytestlatest
+
     - stage: test
     # python x env above are already included into this stage
     - python: "2.7"

--- a/.travis.yml
+++ b/.travis.yml
@@ -21,8 +21,7 @@ env:
 - TOXENV=py-pytest31
 - TOXENV=py-pytest32
 - TOXENV=py-pytest33
-- TOXENV=py-pytest36
-- TOXENV=py-pytest38
+- TOXENV=py-pytestlatest
 
 install: pip install tox setuptools_scm
 script: tox

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -5,8 +5,7 @@ environment:
   - TOXENV: "py34-pytest33"
   - TOXENV: "py35-pytest33"
   - TOXENV: "py36-pytest33"
-  - TOXENV: "py36-pytest36"
-  - TOXENV: "py36-pytest38"
+  - TOXENV: "py36-pytestlatest"
   - TOXENV: "py27-pytest33-pexpect"
   - TOXENV: "py36-pytest33-pexpect"
 

--- a/changelog/349.bugfix.rst
+++ b/changelog/349.bugfix.rst
@@ -1,0 +1,1 @@
+Correctly handle warnings created with arguments that can't be serialized during the transfer from workers to master node.

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -766,6 +766,23 @@ class TestWarnings:
         result = testdir.runpytest(n)
         result.stdout.fnmatch_lines(["*MyWarning*", "*1 passed, 1 warnings*"])
 
+    @pytest.mark.parametrize("n", ["-n0", "-n1"])
+    def test_unserializable_arguments(self, testdir, n):
+        """Check that warnings with unserializable arguments are handled correctly (#349)."""
+        testdir.makepyfile(
+            """
+            import warnings, pytest
+
+            def test_func(tmpdir):
+                fn = (tmpdir / 'foo.txt').ensure(file=1)
+                with fn.open('r') as f:
+                    warnings.warn(UserWarning("foo", f))
+        """
+        )
+        testdir.syspathinsert()
+        result = testdir.runpytest(n)
+        result.stdout.fnmatch_lines(["*UserWarning*foo.txt*", "*1 passed, 1 warnings*"])
+
 
 class TestNodeFailure:
     def test_load_single(self, testdir):

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -400,64 +400,6 @@ class TestTerminalReporting:
             ]
         )
 
-    @pytest.mark.parametrize("n", ["-n0", "-n1"])
-    @pytest.mark.parametrize("warn_type", ["pytest", "builtin"])
-    def test_warnings(self, testdir, n, warn_type):
-        from pkg_resources import parse_version
-
-        if parse_version(pytest.__version__) < parse_version("3.1"):
-            pytest.skip("pytest warnings requires >= 3.1")
-
-        if warn_type == "builtin":
-            warn_code = """warnings.warn(UserWarning('this is a warning'))"""
-        elif warn_type == "pytest":
-            warn_code = """request.config.warn('', 'this is a warning',
-                           fslocation=py.path.local())"""
-        else:
-            assert False
-        testdir.makepyfile(
-            """
-            import warnings, py, pytest
-
-            @pytest.mark.filterwarnings('ignore:config.warn has been deprecated')
-            def test_func(request):
-                {warn_code}
-        """.format(
-                warn_code=warn_code
-            )
-        )
-        result = testdir.runpytest(n)
-        result.stdout.fnmatch_lines(["*this is a warning*", "*1 passed, 1 warnings*"])
-
-    @pytest.mark.parametrize("n", ["-n0", "-n1"])
-    def test_custom_subclass(self, testdir, n):
-        """Check that warning subclasses that don't honor the args attribute don't break
-        pytest-xdist (#344)
-        """
-        from pkg_resources import parse_version
-
-        if parse_version(pytest.__version__) < parse_version("3.1"):
-            pytest.skip("pytest warnings requires >= 3.1")
-
-        testdir.makepyfile(
-            """
-            import warnings, py, pytest
-
-            class MyWarning(UserWarning):
-
-                def __init__(self, p1, p2):
-                    self.p1 = p1
-                    self.p2 = p2
-                    self.args = ()
-
-            def test_func(request):
-                warnings.warn(MyWarning("foo", 1))
-        """
-        )
-        testdir.syspathinsert()
-        result = testdir.runpytest(n)
-        result.stdout.fnmatch_lines(["*MyWarning*", "*1 passed, 1 warnings*"])
-
     def test_logfinish_hook(self, testdir):
         """Ensure the pytest_runtest_logfinish hook is being properly handled"""
         from _pytest import hookspec
@@ -763,6 +705,66 @@ def test_sub_plugins_disabled(testdir, plugin):
     result = testdir.runpytest(p1, "-n1", "-p", "no:%s" % plugin)
     assert result.ret == 0
     result.stdout.fnmatch_lines("*1 passed*")
+
+
+class TestWarnings:
+    @pytest.fixture(autouse=True)
+    def skip_if_unsupported_pytest_version(self):
+        """Skip tests of this class if we are running in a pytest version which does not
+        support warnings yet.
+        """
+        from pkg_resources import parse_version
+
+        if parse_version(pytest.__version__) < parse_version("3.1"):
+            pytest.skip("pytest warnings requires >= 3.1")
+
+    @pytest.mark.parametrize("n", ["-n0", "-n1"])
+    @pytest.mark.parametrize("warn_type", ["pytest", "builtin"])
+    def test_warnings(self, testdir, n, warn_type):
+        if warn_type == "builtin":
+            warn_code = """warnings.warn(UserWarning('this is a warning'))"""
+        elif warn_type == "pytest":
+            warn_code = """request.config.warn('', 'this is a warning',
+                           fslocation=py.path.local())"""
+        else:
+            assert False
+        testdir.makepyfile(
+            """
+            import warnings, py, pytest
+
+            @pytest.mark.filterwarnings('ignore:config.warn has been deprecated')
+            def test_func(request):
+                {warn_code}
+        """.format(
+                warn_code=warn_code
+            )
+        )
+        result = testdir.runpytest(n)
+        result.stdout.fnmatch_lines(["*this is a warning*", "*1 passed, 1 warnings*"])
+
+    @pytest.mark.parametrize("n", ["-n0", "-n1"])
+    def test_custom_subclass(self, testdir, n):
+        """Check that warning subclasses that don't honor the args attribute don't break
+        pytest-xdist (#344)
+        """
+        testdir.makepyfile(
+            """
+            import warnings, py, pytest
+
+            class MyWarning(UserWarning):
+
+                def __init__(self, p1, p2):
+                    self.p1 = p1
+                    self.p2 = p2
+                    self.args = ()
+
+            def test_func(request):
+                warnings.warn(MyWarning("foo", 1))
+        """
+        )
+        testdir.syspathinsert()
+        result = testdir.runpytest(n)
+        result.stdout.fnmatch_lines(["*MyWarning*", "*1 passed, 1 warnings*"])
 
 
 class TestNodeFailure:

--- a/tox.ini
+++ b/tox.ini
@@ -2,7 +2,7 @@
 # if you change the envlist, please update .travis.yml file as well
 envlist=
   linting
-  py{27,34,35,36}-pytest{30,31,32,33,36,38}
+  py{27,34,35,36}-pytest{30,31,32,33,latest}
   py{27,36}-pytest36-pexpect
   py{27,36}-pytest{master,features}
 
@@ -18,8 +18,7 @@ deps =
   pytest31: pytest~=3.1.0
   pytest32: pytest~=3.2.0
   pytest33: pytest~=3.3.0
-  pytest36: pytest~=3.6.0
-  pytest38: pytest~=3.8.0
+  pytestlatest: pytest
   pytestmaster: git+https://github.com/pytest-dev/pytest.git@master
   pytestfeatures: git+https://github.com/pytest-dev/pytest.git@features
   pexpect: pexpect
@@ -27,8 +26,6 @@ deps =
 platform=
   pexpect: linux|darwin
 commands=
-  # always clean to avoid code unmarshal mismatch on old python/pytest
-  py.cleanup -aq
   pytest {posargs}
 
 [testenv:linting]

--- a/xdist/remote.py
+++ b/xdist/remote.py
@@ -12,6 +12,7 @@ import time
 
 import _pytest.hookspec
 import pytest
+from execnet.gateway_base import dumps, DumpError
 
 
 class WorkerInteractor(object):
@@ -181,8 +182,15 @@ def serialize_warning_message(warning_message):
     if isinstance(warning_message.message, Warning):
         message_module = type(warning_message.message).__module__
         message_class_name = type(warning_message.message).__name__
-        message_args = warning_message.message.args
         message_str = str(warning_message.message)
+        # check now if we can serialize the warning arguments (#349)
+        # if not, we will just use the exception message on the master node
+        try:
+            dumps(warning_message.message.args)
+        except DumpError:
+            message_args = None
+        else:
+            message_args = warning_message.message.args
     else:
         message_str = warning_message.message
         message_module = None

--- a/xdist/workermanage.py
+++ b/xdist/workermanage.py
@@ -426,9 +426,16 @@ def unserialize_warning_message(data):
     if data["message_module"]:
         mod = importlib.import_module(data["message_module"])
         cls = getattr(mod, data["message_class_name"])
-        try:
-            message = cls(*data["message_args"])
-        except TypeError:
+        message = None
+        if data["message_args"] is not None:
+            try:
+                message = cls(*data["message_args"])
+            except TypeError:
+                pass
+        if message is None:
+            # could not recreate the original warning instance;
+            # create a generic Warning instance with the original
+            # message at least
             message_text = "{mod}.{cls}: {msg}".format(
                 mod=data["message_module"],
                 cls=data["message_class_name"],

